### PR TITLE
NIFI-6911 Removed the default value for blob property for PutAzureBlobStorage

### PR DESCRIPTION
**Pull request by sushilkm**
#### Description of PR

Fixes bug [NIFI-6911](https://issues.apache.org/jira/browse/NIFI-6911)

Removed default value for Blob property for AzureBlobStorage processors

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by pvillard31** [comment]
Trying to understand the motivation behind this change: is it about user experience? IMO, PutAzureBlobStorage/FetchAzureBlobStorage/DeleteAzureBlobStorage would always rely on incoming flow file attributes to determine the blob, right? Is it just to make the user notice about it and make the user specify its own attribute name for this property?

**Comment by sushilkm** [comment]
A user can provide an expression which generates unique filenames, or a filename could come from other processors via some variable.
Also, the default value it has is currently useless, user has to almost always update that value to something.

**Comment by pvillard31** [comment]
OK, so it's not a bug, it's more about improving the user experience when it comes to the processors. In that case, I'm wondering if it makes sense to only remove the default value for PutAzureBlobStorage/DeleteAzureBlobStorage and not for FetchAzureBlobStorage. The reason is: most of the time we're using the combination of ListAzureBlobStorage with FetchAzureBlobStorage and the List processor is creating flow files with this flow file attribute to be used in the Fetch processor (https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/ListAzureBlobStorage.java#L72). In that case a user would not change the property value. Thoughts?

**Comment by turcsanyip** [comment]
I would agree with @pvillard31. List + Fetch is quite common and List + Delete can also be used.
In case of Put, the upstream processors can use the default attribute name too and in this case no need to configure the property on Put processor.
In terms of user experience, I don't see a big difference between overriding the default value and initialising an empty value.

**Comment by sushilkm** [comment]
I would agree about it is not exactly a bug and it makes sense for List/Fetch combination, however for PutAzureBlobStorage/DeleteAzureBlobStorage it still will make more sense to not have that default, because they are generally not used in conjunction with ListAzureBlobStorage

**Comment by pvillard31** [comment]
Hi @sushilkm - I think the List+Delete makes sense too. I do see value for the change you're proposing with the Put processor. If you update the PR to improve the Put processor, I'm a +1.

**Comment by sushilkm** [comment]
So do you mean that List+Delete is the general usage.
I am not sure how many times would someone want to list+delete instead of deleting a blob based on other conditions or name of blob coming from somewhere else.

**Comment by pvillard31** [comment]
Hey @sushilkm  - sorry for the late reply. I didn't use the Delete processor a lot but I'd say it's pretty common to do something like:
List -> Fetch -> do something on data and send it somewhere else -> Delete

Thoughts?

**Comment by sushilkm** [comment]
@pvillard31  @turcsanyip as suggested I have updated the PR to only update PutAzureBlobStorage processor, please check this one.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureBlobStorage.java [hunk 1]
```diff
@@ -20,9 +20,7 @@
 import java.io.IOException;
 import java.io.InputStream;
 import java.net.URISyntaxException;
-import java.util.HashMap;
-import java.util.Map;
-import java.util.Optional;
+import java.util.*;
```
Comment (jfrazee) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureBlobStorage.java:23]: Minor thing: I'd suggest not using a wildcard import here.

Comment (sushilkm) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureBlobStorage.java]: The IDE did automatic imports, I have not fixed them

Comment (sushilkm) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureBlobStorage.java]: The IDE did automatic imports, I have now fixed them

**Comment by MuazmaZ** [comment]
@sushilkm it looks good to me. Tested the functionality and it changes the default so the user can enter the blob name.

**Comment by jfrazee** [comment]
@sushilkm What will the behavior be when we upgrade NiFi with an existing flow? I assume all the Puts are going to have to be reconfigured. If so will need to update the migration guidance.

**Comment by joewitt** [comment]
We should avoid relying on manual migration steps for changes in processors where possible.  Is it not possible here?

**Comment by sushilkm** [comment]
@joewitt @jfrazee since it is using the same property name 'blob', it would work as is with the upgrades, users will not require any steps towards upgrades, I tested it too.

**Comment by pvillard31** [comment]
When the processor is added to the canvas the default value is written as the value of the property of the flow.xml.gz and it's also stored in the NiFi Registry. I believe that removing default values won't affect existing deployments. There would be an impact if there is a tool automatically building flows from scratch that would rely on this default value. But this can be documented I think. Did you think about another scenario @jfrazee ?

**Comment by sushilkm** [comment]
@pvillard31 I am ok to document this behavior, any pointers to where do I need to update this piece of information

**Comment by jfrazee** [comment]
@pvillard31 @sushilkm That covers everything. Tried the different scenarios and LGTM.

**Comment by turcsanyip** [comment]
My opinion still is that removing the default value is not a good idea. It is quite common that Put file/blob object processors have the core attribute `${filename}` as their default value. It exists on all FlowFiles so it cannot happen that it has no value (unlike `${azure.blobname}` which is not a standard attribute).

The other Azure Blob processors could also be modified to use this `${filename}` (while preserving the legacy `${azure.blobname}` as well for List/Fetch in order to maintain backward compatibility). In this way, the Azure Blob processors would be standardized to the `${filename}` convention. Furthermore, the property descriptor overriding (BLOB vs BLOB_NAME) would not be needed because all of them would use the same default value.

**Comment by jfrazee** [comment]
@turcsanyip I was thinking that sounded like a really easy, elegant solution, but it looks like `${filename}` is the basename whereas `${azure.blobname}` isn't if there are any "directories". I think there's probably still a viable path to migrating the processors to `${filename}` but I think it'd require more code than this PR. That said, using `${filename}`, `${path}`, etc. would bring it closer to the other filesystem processors.

**Comment by turcsanyip** [comment]
@jfrazee I meant the `${filename}` would contain the whole object name, including the pseudo directory name too. So no path / filename separation.
In this way, the Azure Blob processors would work the same way as the S3 and GCS processors: the List processor fills in `${filename}` on the outgoing FlowFile, the others (Fetch, Put, Delete) use `${filename}` as default value.

If someone would like to handle / configure the directory and the filename separately, then the new AzureDataLakeStorage processors would be better choice. I would keep the single "Blob" property for the Blob processors.

**Comment by jfrazee** [comment]
@turcsanyip More details below but the tl;dr is that I think this makes sense, but `${filename}` doesn't currently contain the whole object name so I think we'd have to break `ListAzureBlobStorage` to make the new default useful.

E.g., if you do a recursive listing and have `/test/test.txt` what you end up with is:

```properties
azure.blobname = test/test.txt
filename = test.txt
path = ./
```

So if `FetchAzureBlobStorage` defaulted to `${filename}`, I think `ListAzureBlobStorage` => `FetchAzureBlobStorage` will break unless we change how `${filename}` is defined in `ListAzureBlobStorage`. But that will break things for anyone currently using `${filename}` with the Blob processors.

Also, arguably, `${path}` is wrong and there's a bug here because that's not the object's path. Comparing all this to `ListFile`, the behavior is different:

```properties
absolute.path = /tmp/nifi/test/
filename = test.txt
path = test/
```

S3 and GCS might be unique in the way they treat `${filename}`. A lot of processors set `${filename}` to a bare filename. And the other filesystem-oriented `Fetch` processors, if they have a default use `${absolute.path}/${filename}` or `${path}/${filename}`.

Options:
1. Go with this PR.
2. Go for consistency with S3 and GCS, change `${filename}` to be the full object name, use it as the default, potentially breaking flows that assume it's the bare filename today.
3. Go for consistency with File, fix `${path}` and default to `${path}/${filename}`, potentially breaking flows using `${path}` today. But `${path}` is wrong, so is anyone using it?
4. Do nothing.

**Comment by turcsanyip** [comment]
@jfrazee Thanks for listing the options and also for pointing out that the `${filename}` solution (option 2) might break existing flows. So it is not backward compatible.
I feel option 3 a bit complicated.
Option 1 vs 4: I would rather vote for 4 (that is leaving it as it is now), but we can go with this PR and it can be merged if you like this solution.

**Pull request closed without merging**